### PR TITLE
fix: end build sessions before launching cargo applications

### PR DESCRIPTION
**Pull request by jdx**
`mbx run` currently keeps the cache session alive for the application's lifetime. Builds started by that application inherit its socket, compiler adapters, roots, and policy; their mbx diagnostics can appear on the launcher's terminal even when the application captures subprocess output.

This change uses an internal Cargo runner to capture Cargo's selected executable, arguments, working directory, and launch environment. After Cargo returns, mbx commits and closes the build session, prints its cache summary, and starts the application through the caller's configured runner. Only mbx's environment overrides are restored, including the original PATH and compiler wrappers; Cargo's dynamic-library paths remain intact. The application's exit status is returned to the caller. Automatic garbage collection and lifetime-savings reporting remain after application exit, so a target sweep cannot remove the executable before launch.

Sessions carry an environment restoration record and an owner lock. A subsequent mbx/Cargo-shim invocation can restore the original settings and establish a new session after that owner exits. Live nested build orchestration continues to reuse its session. The artifact store and machine-wide scheduler remain shared.

Scope and compatibility:
- Covers direct `run` and `r` invocations. Arbitrary shell aliases and application launches inside external orchestration tools still require a separate boundary design.
- Existing config-file and environment runners are resolved with `cargo-config2`, including target-specific runners and runner arguments. Launches using `+toolchain`, `--config`, or `-C` use ordinary Cargo with the caller's environment because the resolver cannot faithfully resolve those overrides. Those invocations retain Cargo behavior but do not receive an mbx build session.
- Stale-session recovery uses restoration data emitted by this version; it cannot reconstruct overwritten caller values from older mbx sessions.

Validation: `mise run ci` passes, including debug/release Clippy, workspace tests, all 59 Bats cases, release diagnostics, generated documentation, and documentation links. New integration cases cover environment and exit-status restoration, configured runners/wrappers, a detached daemon capturing a build in another workspace after the launcher exits, stale-session recovery, live watch/nextest-style nesting, command-line runner overrides, and non-Unicode environment values with temporary paths containing spaces.

Addresses https://github.com/jdx/mr-boxington/discussions/434.

---

> [!NOTE]
> **Medium Risk**
> Changes process lifecycle and environment restoration for `cargo run`, with complex PATH/runner handling; mistakes could leak session state or break launches, but behavior is heavily covered by new integration tests.
> 
> **Overview**
> **`cargo run` / `r` no longer keeps the mbx cache session alive while the built binary runs.** A new launch path installs a temporary `mbx-launch` Cargo runner shim that records the executable, args, cwd, and launch environment; after Cargo succeeds, mbx shuts down the session (including cache summary) and starts the app with the caller’s restored env and configured target runner, returning the application’s exit code.
> 
> **Session boundaries and recovery:** Build sessions record env overlays (`MBX_SESSION_RESTORE`) and hold an owner lock (`MBX_SESSION_LEASE`) so later invocations can restore the original PATH/wrappers when the launcher exits, while live nested orchestrators (e.g. watch/nextest) still reuse an active session. **`+toolchain`, `--config`, and `-C` on `run`** bypass this path and invoke plain Cargo with the remembered caller environment (no mbx session) because `cargo-config2` cannot resolve those overrides safely.
> 
> Shared **`cargo_subcommand`** parsing replaces duplicated logic in the cargo shim. Adds **`cargo-config2`** and integration tests in `session_scope.rs` for env restoration, runners, detached builds, stale recovery, and edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 43a2be1fb54ebb9a13e1c50e992bfc751030ba96. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: crates/mbx/src/cli/launch.rs [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+        .take_while(|arg| arg.as_str() != "--")
+        .collect();
+    args.iter().any(|arg| matches!(arg.as_str(), "run" | "r"))
+        && args.iter().any(|arg| {
+            arg.starts_with('+')
+                || arg.as_str() == "--config"
+                || arg.starts_with("--config=")
+                || arg.as_str() == "-C"
+        })
+}
```
[the remark this answers is not in the document]
Comment (jdx) [reply crates/mbx/src/cli/launch.rs:90]: Fixed in 43a2be1. Plain-launch detection now uses the same Cargo-aware subcommand parser as capture, recovery, and the Cargo shim, so run/r values after test, build, or nextest are not classified as launches. Added regression cases for test run, --bin run, --features run, and global option values.

File: crates/mbx/src/cli/launch.rs [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+            .any(|arg| matches!(arg.as_str(), "--help" | "-h"))
+        {
+            return Ok(None);
+        }
+        let command = args
+            .iter()
+            .find(|arg| !arg.starts_with('+') && !arg.starts_with('-'));
+        if !command.is_some_and(|arg| matches!(arg.as_str(), "run" | "r")) {
+            return Ok(None);
+        }
```
[the remark this answers is not in the document]
Comment (jdx) [reply crates/mbx/src/cli/launch.rs:199]: Fixed in 43a2be1 with one shared parser that skips value-taking global options. The integration test now enters through Cargo shim mode with --color never run and verifies that the application receives no build-session overlay.

File: crates/mbx/src/cli/launch.rs [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+            restore.insert(name.clone(), caller.get(&name).cloned());
+        }
+    }
+    for name in environment
+        .keys()
+        .map(OsString::from)
+        .chain(["PATH", "CARGO"].into_iter().map(OsString::from))
+    {
+        restore.insert(name.clone(), caller.get(&name).cloned());
+    }
```
[the remark this answers is not in the document]
Comment (jdx) [reply crates/mbx/src/cli/launch.rs:150]: Fixed in 43a2be1. Application capture retains Cargo's CARGO value and runtime PATH prefix while replacing only the inherited build-PATH suffix with the caller's PATH. Added prefix/suffix regression coverage and run the application integration fixture with -C prefer-dynamic. Stale CLI recovery still restores the original caller environment.

File: crates/mbx/src/cli/launch.rs [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+            .iter()
+            .any(|arg| matches!(arg.as_str(), "--help" | "-h"))
+        {
+            return Ok(None);
+        }
+        let command = args
+            .iter()
+            .find(|arg| !arg.starts_with('+') && !arg.starts_with('-'));
+        if !command.is_some_and(|arg| matches!(arg.as_str(), "run" | "r")) {
+            return Ok(None);
```
[the remark this answers is not in the document]
Comment (jdx) [reply crates/mbx/src/cli/launch.rs]: Fixed in 43a2be1. Capture, stale/nested recovery, plain-launch classification, and Cargo shim routing now share a parser that skips global-option values. The integration fixture uses --color never run through Cargo shim mode. Full mise run ci and all seven session-scope integration tests pass locally.

File: crates/mbx/src/cli/launch.rs [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+
+/// Config overrides cannot be resolved by cargo-config2. Let Cargo handle
+/// these launches with the caller's environment until it offers a stable
+/// resolved-config API. In particular, never replace an unknown runner.
+pub(super) fn needs_plain_launch(arguments: &[String]) -> bool {
+    let args: Vec<_> = arguments
+        .iter()
+        .take_while(|arg| arg.as_str() != "--")
+        .collect();
+    args.iter().any(|arg| matches!(arg.as_str(), "run" | "r"))
```
[the remark this answers is not in the document]
Comment (jdx) [reply crates/mbx/src/cli/launch.rs]: Fixed in 43a2be1 with a shared Cargo-aware parser used by prepare, recovery, fallback classification, and the Cargo shim. Regression tests cover run as a test/bin/feature name and global-option values; the integration fixture exercises --color never run. Full mise run ci passes.

File: crates/mbx/tests/session_scope.rs [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+    for _ in 0..200 {
+        if result.exists() {
+            break;
+        }
+        std::thread::sleep(std::time::Duration::from_millis(50));
+    }
+    assert_eq!(
+        std::fs::read_to_string(result).unwrap(),
+        "independent build captured"
+    );
```
[the remark this answers is not in the document]
Comment (jdx) [reply crates/mbx/tests/session_scope.rs]: Fixed in 43a2be1: the loop now polls for the complete expected content with the same retry count and delay. All seven session-scope integration tests pass after this change.
